### ISSUE TITLE
python bindings: set_alert_notify may deadlock when called more than once

### ISSUE THREAD

**Issue by WGH-** [arvidn/libtorrent#4638]
libtorrent version (or branch): 1.2.3 (the relevant part appears to not have changed since)

platform/architecture: Linux/amd64

compiler and compiler version: gcc 9.2.0

If you call `set_alert_notify` more than once, it may deadlock. Here are the relevant backtraces:
```
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x7fa5251b2300, clockid=<optimized out>, expected=0, futex_word=0x7fa5275b51fc <_PyRuntime+1340>) at ../sysdeps/unix/sysv/linux/futex-internal.h:208
#1  __pthread_cond_wait_common (abstime=0x7fa5251b2300, clockid=<optimized out>, mutex=0x7fa5275b5200 <_PyRuntime+1344>, cond=0x7fa5275b51d0 <_PyRuntime+1296>) at pthread_cond_wait.c:520
#2  __pthread_cond_timedwait (cond=0x7fa5275b51d0 <_PyRuntime+1296>, mutex=0x7fa5275b5200 <_PyRuntime+1344>, abstime=0x7fa5251b2300) at pthread_cond_wait.c:656
#3  0x00007fa5273e2d3b in take_gil () from /usr/lib64/libpython3.7m.so.1.0
#4  0x00007fa5273e31cf in PyEval_RestoreThread () from /usr/lib64/libpython3.7m.so.1.0
#5  0x00007fa5273c0a66 in PyGILState_Ensure () from /usr/lib64/libpython3.7m.so.1.0
#6  0x00007fa525f3f58c in lock_gil::lock_gil (this=<synthetic pointer>) at /usr/src/debug/net-libs/libtorrent-rasterbar-1.2.3/libtorrent-libtorrent-1_2_3-python3_7/bindings/python/src/gil.hpp:29
#7  (anonymous namespace)::alert_notify (cb=...) at /usr/src/debug/net-libs/libtorrent-rasterbar-1.2.3/libtorrent-libtorrent-1_2_3-python3_7/bindings/python/src/session.cpp:394
#8  0x00007fa525f5bef2 in std::__invoke_impl<void, void (*&)(boost::python::api::object), boost::python::api::object&> (__f=<optimized out>) at /usr/include/boost/python/refcount.hpp:14
#9  std::__invoke<void (*&)(boost::python::api::object), boost::python::api::object&> (__fn=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/bits/invoke.h:95
#10 std::_Bind<void (*(boost::python::api::object))(boost::python::api::object)>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) (__args=..., this=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/functional:400
#11 std::_Bind<void (*(boost::python::api::object))(boost::python::api::object)>::operator()<, void>() (this=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/functional:484
#12 std::_Function_handler<void (), std::_Bind<void (*(boost::python::api::object))(boost::python::api::object)> >::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/bits/std_function.h:300
#13 0x00007fa5256fce17 in std::function<void ()>::operator()() const (this=0x562f149dfb18) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/bits/std_function.h:685
#14 libtorrent::alert_manager::maybe_notify (this=this@entry=0x562f149dfa98, a=a@entry=0x7fa520003030) at alert_manager.cpp:73
```

```
#0  __lll_lock_wait (futex=futex@entry=0x562f149dfa98, private=0) at lowlevellock.c:52
#1  0x00007fa5272c8881 in __GI___pthread_mutex_lock (mutex=0x562f149dfa98) at ../nptl/pthread_mutex_lock.c:115
#2  0x00007fa5256fceca in __gthread_mutex_lock (__mutex=0x562f149dfa98) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/x86_64-pc-linux-gnu/bits/gthr-default.h:749
#3  __gthread_recursive_mutex_lock (__mutex=0x562f149dfa98) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/x86_64-pc-linux-gnu/bits/gthr-default.h:811
#4  std::recursive_mutex::lock (this=0x562f149dfa98) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/mutex:106
#5  std::unique_lock<std::recursive_mutex>::lock (this=0x7ffdcd1f2f00, this=0x7ffdcd1f2f00) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/bits/unique_lock.h:141
#6  std::unique_lock<std::recursive_mutex>::unique_lock (__m=..., this=0x7ffdcd1f2f00) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/g++-v9/bits/unique_lock.h:71
#7  libtorrent::alert_manager::set_notify_function(std::function<void ()> const&) (this=0x562f149dfa98, fun=...) at alert_manager.cpp:90
#8  0x00007fa52587ff5e in libtorrent::session_handle::set_alert_notify(std::function<void ()> const&) (this=this@entry=0x562f1498da90, fun=...) at ../include/libtorrent/aux_/session_impl.hpp:642
#9  0x00007fa525f3efb6 in (anonymous namespace)::set_alert_notify (s=..., cb=...) at /usr/include/boost/python/refcount.hpp:14
```

Here thread 1 attempts to call the callback, and is locked on attempting to take Python GIL, while holding `alert_manager`'s mutex. Thread 2 attempts to change the current callback, and is locked on attempting to take `alert_manager`'s mutex, while holding Python GIL.

**Comment by arvidn** [comment]
I believe the most reasonable solution is to mark `set_alert_notify` to not hold the GIL. Does that sound reasonable?

**Comment by arvidn** [comment]
I think it's called `allow_threads` in boost.python

**Comment by WGH-** [comment]
It sounds reasonable. But I don't know how Boost handles Python objects. Is it safe to store (copy? move?) it into `std::bind` object?

**Comment by WGH-** [comment]
I'll try it soon.

**Comment by WGH-** [comment]
`set_alert_notify` is not only function that is prone to deadlocks. I think any function that doesn't release the GIL will likely deadlock when `set_alert_notify` callback is set. And unlike `set_alert_notify`, there is no workaround "just call it only once", unfortunately.

For example, I noticed deadlocks with `get_torrent_status` as well.

```
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x7f71b03a0970, clockid=<optimized out>, expected=0, futex_word=0x7f71b3d9d1f8 <_PyRuntime+1336>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:208
#1  __pthread_cond_wait_common (abstime=0x7f71b03a0970, clockid=<optimized out>, mutex=0x7f71b3d9d200 <_PyRuntime+1344>, cond=0x7f71b3d9d1d0 <_PyRuntime+1296>)
    at pthread_cond_wait.c:520
#2  __pthread_cond_timedwait (cond=0x7f71b3d9d1d0 <_PyRuntime+1296>, mutex=0x7f71b3d9d200 <_PyRuntime+1344>, abstime=0x7f71b03a0970) at pthread_cond_wait.c:656
#3  0x00007f71b3bcaceb in take_gil () from /usr/lib64/libpython3.7m.so.1.0
#4  0x00007f71b3bcb17f in PyEval_RestoreThread () from /usr/lib64/libpython3.7m.so.1.0
#5  0x00007f71b3ba8a46 in PyGILState_Ensure () from /usr/lib64/libpython3.7m.so.1.0
#6  0x00007f71b272b5da in lock_gil::lock_gil (this=<synthetic pointer>)
    at /usr/src/debug/net-libs/libtorrent-rasterbar-1.2.3/libtorrent-libtorrent-1_2_3-python3_7/bindings/python/src/gil.hpp:29
#7  (anonymous namespace)::alert_notify (cb=...)
    at /usr/src/debug/net-libs/libtorrent-rasterbar-1.2.3/libtorrent-libtorrent-1_2_3-python3_7/bindings/python/src/session.cpp:394
#8  0x00007f71b2747f40 in std::__invoke_impl<void, void (*&)(boost::python::api::object), boost::python::api::object&> (__f=<optimized out>)
    at /usr/include/boost/python/refcount.hpp:14
#9  std::__invoke<void (*&)(boost::python::api::object), boost::python::api::object&> (__fn=<optimized out>)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/invoke.h:95
#10 std::_Bind<void (*(boost::python::api::object))(boost::python::api::object)>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) (__args=...,
    this=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/functional:400
#11 std::_Bind<void (*(boost::python::api::object))(boost::python::api::object)>::operator()<, void>() (this=<optimized out>)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/functional:484
#12 std::_Function_handler<void (), std::_Bind<void (*(boost::python::api::object))(boost::python::api::object)> >::_M_invoke(std::_Any_data const&) (__functor=...)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/std_function.h:300
#13 0x00007f71b1ee8e17 in std::function<void ()>::operator()() const (this=0x55a14f4ec4e8)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/std_function.h:683
#14 libtorrent::alert_manager::maybe_notify (this=this@entry=0x55a14f4ec468, a=a@entry=0x7f71a8252820) at alert_manager.cpp:73
```

```
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0x55a14f4eb508) at ../sysdeps/unix/sysv/linux/futex-internal.h:80
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x55a14f4eb4b8, cond=0x55a14f4eb4e0) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x55a14f4eb4e0, mutex=0x55a14f4eb4b8) at pthread_cond_wait.c:638
#3  0x00007f71b1b4a51c in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/libstdc++.so.6
#4  0x00007f71b206b69b in libtorrent::aux::torrent_wait (done=@0x7fffddcc6fff: false, ses=...) at session_call.cpp:77
#5  0x00007f71b207811c in libtorrent::session_handle::sync_call<void (libtorrent::aux::session_impl::*)(std::vector<libtorrent::torrent_status, std::allocator<libtorrent::torrent_status> >*, std::function<bool (libtorrent::torrent_status const&)> const&, libtorrent::flags::bitfield_flag<unsigned int, libtorrent::status_flags_tag, void>) const, std::vector<libtorrent::torrent_status, std::allocator<libtorrent::torrent_status> >*, std::function<bool (libtorrent::torrent_status const&)> const&, libtorrent::flags::bitfield_flag<unsigned int, libtorrent::status_flags_tag, void> const&>(void (libtorrent::aux::session_impl::*)(std::vector<libtorrent::torrent_status, std::allocator<libtorrent::torrent_status> >*, std::function<bool (libtorrent::torrent_status const&)> const&, libtorrent::flags::bitfield_flag<unsigned int, libtorrent::status_flags_tag, void>) const, std::vector<libtorrent::torrent_status, std::allocator<libtorrent::torrent_status> >*&&, std::function<bool (libtorrent::torrent_status const&)> const&, libtorrent::flags::bitfield_flag<unsigned int, libtorrent::status_flags_tag, void> const&) const (this=this@entry=0x55a14f3e3280, f=<optimized out>)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/shared_ptr_base.h:1020
#6  0x00007f71b206c75e in libtorrent::session_handle::get_torrent_status(std::function<bool (libtorrent::torrent_status const&)> const&, libtorrent::flags::bitfield_flag<unsigned int, libtorrent::status_flags_tag, void>) const (this=this@entry=0x55a14f3e3280, pred=..., flags=flags@entry=...) at session_handle.cpp:187
#7  0x00007f71b272c8e1 in (anonymous namespace)::get_torrent_status (s=..., pred=..., flags=0) at /usr/include/boost/python/refcount.hpp:14
```

**Comment by arvidn** [comment]
yeah, I think the fundamental problem is that python code cannot be allowed to be run directly while blocked in another call (which may hold the GIL already).

I think a proper solution to this would have to take some liberties in how `set_alert_notify` is exposed to python. To be fair, it *is* a very low level function that's really just there to allow integration with a high level message queue.

Either, the notify function is wrapped in the python binding, and the function that's actually set enqueues a call to the python function somehow. To be deferred and run, say, when the GIL is released.

I don't know enough about python to know whether there is such a feature or not.

Alternatively, we could raise the abstraction level and instead make the python API take some kind of message queue to post a message to, rather than calling a callback.

**Comment by WGH-** [comment]
The easy solution available on UNIX-like systems it to return some file descriptor (backed by linux-specific eventfd or just plain pipe) that can be integrated into external event loop: when libtorrent wants to signal something, it simply does a non-blocking write to that FD.

Not sure how to extend that to Windows.

**Comment by AllSeeingEyeTolledEweSew** [comment]
@WGH- could your application live with `wait_for_alert()` instead? See also my feature request in #4668

**Comment by WGH-** [comment]
> @WGH- could your application live with `wait_for_alert()` instead? See also my feature request in #4668

I think yes. There're some problems related to this approach, which you outlined in your issue, though.

**Comment by arvidn** [comment]
python has an abstraction over pipes, would that be a reasonable API in python?

**Comment by WGH-** [comment]
> python has an abstraction over pipes, would that be a reasonable API in python?

Which abstraction are you referring to?

**Comment by AllSeeingEyeTolledEweSew** [comment]
It appears windows supports AF_UNIX in recent builds.

The socket-ish approach has the cool feature that an application can just add the handle to its `select()` call. Potentially reduces the number of event handling threads needed, especially in cpython which doesn't have parallelism anyway. Even C++ applications could use it.

FWIW, the only other way I'm aware of to schedule python code to run is with something like the existing `wait_for_alert()`. Python doesn't have a native message loop any more than C++ does. Every synchronization primitive I know about (everything in `threading` and `queue`) relies on python code, so you'd need to take the GIL to modify it.

**Comment by WGH-** [comment]
If we're going to go the fd route, we should check how Python does this for `signal.set_wakeup_fd`, because it is exactly the same pattern, and it even claims Windows support. Our case is even slightly easier, as we don't have to do that from the signal context.

Specifically, see 
https://github.com/python/cpython/blob/a59fc9160db50a5e20358303e7ddbda180a1b74f/Modules/signalmodule.c#L289-L327
https://github.com/python/cpython/blob/a59fc9160db50a5e20358303e7ddbda180a1b74f/Modules/signalmodule.c#L665-L775

**Comment by arvidn** [comment]
Here's a strawman:

deprecate the `set_alert_notify()` function in the python binding. Instead expose something like: `get_alert_fd()` which returns a file-like object that can be used in select. Either an `AF_UNIX` socket, or one end of an `os.pipe()`. The binding can then write a byte to the other end of that socket pair/pipe every time an alert is posted.

Ideally it would only write a byte when the first alert is posted with an empty queue, but that might be a bit tricky with the current API.

Ideally it would be an `eventfd`, but there doesn't seem to be a platform independent abstraction for that in python.

**Comment by WGH-** [comment]
I still wonder why `signal.set_wakeup_fd` accepts a write end of a FD instead of calling `pipe` itself and returning the read end. On Linux and like, there's little point in using anythng other than pipe. 

On Windows, however, one can choose either `os.pipe` or socket. The benefits of the later are obvious to me: it cleanly integrates into your typical socket event loop. Why would anyone want to use `os.pipe` on Windows is not clear to me. I even found a Python issue which discussed the socket support for `set_wakeup_fd` on Windows (https://bugs.python.org/issue22018), and it seems that there's indeed little point of using pipes here. The only reason it wasn't even removed there is to not cause a regression.

> Ideally it would be an `eventfd`, but there doesn't seem to be a platform independent abstraction for that in python.

Although if was me who initially mentioned `eventfd`, I'd say it only provides marginal benefit over pipes in this scenario. It pleases perfectionists and people who like new and shiny stuff. Pipes are actually perfectly fine here.

**Comment by AllSeeingEyeTolledEweSew** [comment]
`set_alert_fd()` lets the application decide which one it wants to do in a straightforward way, rather than a `get_alert_fd()` restricting to one return type, or supporting extra options to configure the returned fd. My two cents is libtorrent should implement `set` over `get`.

**Comment by AllSeeingEyeTolledEweSew** [comment]
@WGH- maybe academic, but is there a security benefit of using pipes over sockets? If I make an `AF_INET` socket for event signaling on windows (on versions where `AF_UNIX` isn't supported), maybe an XSS'ed site could send garbage to it? I skimmed that issue but didn't see security mentioned.

**Comment by arvidn** [comment]
sure, a pipe or a socket will probably work fine in most cases. But keep in mind that if all alerts are enabled, you can get bursts of 10000 alerts before you have a chance to react and pop them. That's 10 kB of dummy bytes transported through the kernel. If you're unlucky, you hit the pipe buffer size and the alert notify function blocks.

You're saying you can't use a pipe fd and a socket in the same select() call on windows? Typically you can put all kinds of `HANDLE`s in `WaitForMultipleObjects()`

**Comment by arvidn** [comment]
@AllSeeingEyeTolledEweSew I think there might be a security issue on a machine with multiple users. Where another user may have access to your socket.

Regarding `set_` versus `get_`, that makes sense. However, I think one important feature is that whatever the user passes in, can be "signalled" from C++ without having to acquire the GIL. So, probably not signally by calling some `Py_*` function.

**Comment by WGH-** [comment]
>  If I make an AF_INET socket for event signaling on windows (on versions where AF_UNIX isn't supported), maybe an XSS'ed site could send garbage to it? I skimmed that issue but didn't see security mentioned.

`socketpair` returns an already connected pair of sockets, and you can't create an additional connection to any of them. On Windows, `socket.socketpair` is implemented by creating a listening socket and immediately connecting to it: https://github.com/python/cpython/blob/b831129123dbb4bfba49824ae893448b05398f27/Lib/socket.py#L614-L649 In order to interfere with it, not you have to guess the port, but also get into small window between `listen` and `connect`.

> sure, a pipe or a socket will probably work fine in most cases. But keep in mind that if all alerts are enabled, you can get bursts of 10000 alerts before you have a chance to react and pop them. That's 10 kB of dummy bytes transported through the kernel. If you're unlucky, you hit the pipe buffer size and the alert notify function blocks.

I thought libtorrent creates only one notification per batch of new alerts.

Besides, notification fd should be in non-blocking mode all the time anyway.

> You're saying you can't use a pipe fd and a socket in the same select() call on windows? Typically you can put all kinds of HANDLEs in WaitForMultipleObjects()

As far as I know, yes. Mostly because `select` actually comes from the socket library, which is somewhat isolated from other Windows APIs.

> Regarding set_ versus get_, that makes sense. However, I think one important feature is that whatever the user passes in, can be "signalled" from C++ without having to acquire the GIL. So, probably not signalled by calling some Py_* function.

I think it should just accept just plain integer file descriptors. In fact, I think it might be useful to expose this API to C++, because users might want to integrate it into some event loop.

**Comment by AllSeeingEyeTolledEweSew** [comment]
>I think one important feature is that whatever the user passes in, can be "signalled" from C++ without having to acquire the GIL. So, probably not signalled by calling some `Py_*` function.

Yep, that's satisfied here. You'd just call `send()` in C++, and the functions on the python side are just wrapping `recv()`. (or `write()` / `read()`, or whatever)

> sure, a pipe or a socket will probably work fine in most cases. But keep in mind that if all alerts are enabled, you can get bursts of 10000 alerts before you have a chance to react and pop them. That's 10 kB of dummy bytes transported through the kernel. If you're unlucky, you hit the pipe buffer size and the alert notify function blocks.

[set_wakeup_fd(warn_on_buffer_full=False)](https://docs.python.org/3.8/library/signal.html#signal.set_wakeup_fd) addresses this in a pretty general way.

Libtorrent's `set_alert_fd()` could just have the behavior that `set_notify_alert()` currently has, such that you only write a byte when the alert queue goes from 0 to 1. Then the buffer size should be irrelevant I think. For my part I would want behavior like this.

> You're saying you can't use a pipe fd and a socket in the same select() call on windows? Typically you can put all kinds of `HANDLE`s in `WaitForMultipleObjects()`

[The select documentation](https://docs.python.org/3/library/select.html#select.select) says "File objects on Windows are not acceptable, but sockets are. On Windows, the underlying `select()` function is provided by the WinSock library, and does not handle file descriptors that don’t originate from WinSock."

I don't know the background on why they wouldn't use `WaitForMultipleObjects()` over `select()`, it seems a better fit from what you say. I do see there are examples of people calling `WaitForMultipleObjects()` via `ctypes`, but I can't find an example that seems really portable.

In any case, it seems that we could start add `set_alert_fd(int)` and `set_alert_fd(HANDLE)`, with restrictions that are good enough for now (`HANDLE`s are restricted to sockets and pipes only?), and later we can loosen restrictions without ABI changes.

**Comment by arvidn** [comment]
> I thought libtorrent creates only one notification per batch of new alerts.

You're right. I forgot about that.

> In any case, it seems that we could start add set_alert_fd(int) and set_alert_fd(HANDLE), with restrictions that are good enough for now

Are you suggesting not having a platform independent API for this, in python?
Does python have a `HANDLE` type?
Would the handle be assumed to be a socket or an event? It can be many things.

As far as I can tell, both files and sockets have a `fileno()` function in python, suggesting `set_alert_fd(int)` could be sufficient. Interestingly, pipes don't seem to have this function (but it's possible I'm misreading this).

**Comment by arvidn** [comment]
how does this look? https://github.com/arvidn/libtorrent/pull/4672

There should probably be a test for it as well

**Comment by WGH-** [comment]
>  Interestingly, pipes don't seem to have this function (but it's possible I'm misreading this).

I still don't know what are you referring to. As far as I know, there's only `os.pipe`, and no higher-level abstraction.

**Comment by arvidn** [comment]
> I still don't know what are you referring to. As far as I know, there's only os.pipe, and no higher-level abstraction.

I meant it's an abstraction on top of the OS facility. you get (I imagine) a python File I/O object instead of an `int`

**Comment by WGH-** [comment]
> I meant it's an abstraction on top of the OS facility. you get (I imagine) a python File I/O object instead of an int

No, `os.pipe` returns a pair of plain integer file descriptors. Most functions in the `os` module either take or return integer fds, actually.

**Comment by arvidn** [comment]
I see

**Comment by AllSeeingEyeTolledEweSew** [comment]
> No, `os.pipe` returns a pair of plain integer file descriptors. Most functions in the `os` module either take or return integer fds, actually.

https://github.com/python/cpython/blob/master/Modules/_winapi.c#L715

It looks like `os.pipe()`'s return values are cast form `HANDLE`, actually.

> Are you suggesting not having a platform independent API for this, in python?

I actually imagined you might implement `set_alert_fd()` in the C++ API, but I think you're right to just do it in the python wrapper.

> Does python have a `HANDLE` type?
> Would the handle be assumed to be a socket or an event? It can be many things.

Python doesn't really have a `HANDLE` type. 

There is a general pattern that objects have a `fileno()` method if they are backed by some object with a file descriptor. Some system call wrappers honor this, so you can call `select(myObj)` if `myObj.fileno()` returns a file descriptor appropriate for `select()` on your platform. For sockets created on windows, `s.fileno()` is actually a `SOCKET` cast to a 64-bit int.

As I mentioned in the PR, `signal.set_wakeup_fd()` tries to test if the integer argument was a `SOCKET` or a small integer file descriptor.

I am not sure what the deal is with `os.pipe()` on windows. This is the only other place I can find in cpython where `HANDLE` is cast to integer. It's unclear to me that `os.read()` will actually work on such pipes (I think the codepath expects C `int` internally); maybe you need `_winapi.ReadFile()`. The multiprocessing library actually does this. I don't have a windows box to test, I'll have to find a VM.

**Comment by ichorid** [comment]
Could this problem affect earlier versions of Libtorrent? That would explain a lot for problems we witnessed before.

**Comment by arvidn** [comment]
yes, I think any python program using `set_alert_notify()` is at risk.

**Comment by ichorid** [comment]
Is there a kind of workaround for the issue? Could we, say, add some Python-level locking to prevent the problem? Also, is there an easy way to trigger this to build a Python-based unit test?

**Comment by arvidn** [comment]
polling alerts instead of relying on `set_alert_notify` is one work-around.
This patch is another: https://github.com/arvidn/libtorrent/pull/4672 (but it doesn't seem to work on windows yet, I think I need to trouble shoot this on windows in person).

> Also, is there an easy way to trigger this to build a Python-based unit test?

I don't know what you mean. You mean a unit test to trigger the deadlock?

**Comment by ichorid** [comment]
>I don't know what you mean. You mean a unit test to trigger the deadlock?
Yes, exactly. If we're going to build a workaround, we'll have to first produce a (failing) unit test that triggers the deadlock.

**Comment by arvidn** [comment]
I think the `set_alert_notify()` API is inherently problematic in python, so the work-around is to deprecate that function and instead introduce `set_alert_fd()`

**Comment by WGH-** [comment]
> Yes, exactly. If we're going to build a workaround, we'll have to first produce a (failing) unit test that triggers the deadlock.

In my case, I had lots of alerts enabled, and `set_alert_notify` called every iteration. Deadlock happens very soon, you don't even have to have any torrents added, DHT activity will trigger it.

```python
def get_alert_mask():
    category_t = lt.alert.category_t

    return (
          category_t.status_notification
        | category_t.error_notification
        | category_t.dht_log_notification
        | category_t.port_mapping_notification
        | category_t.session_log_notification
    )

async def await_for_alert(ses: lt.session):
    loop = asyncio.get_event_loop()
    event = asyncio.Event()

    def cb():
        loop.call_soon_threadsafe(lambda: event.set())

    ses.set_alert_notify(cb)

    while True:
        event.clear()
        alerts = ses.pop_alerts()
        if alerts:
            return alerts
        await event.wait()

async def main():
    ...
    while True:
        for events in await await_for_alert(ses):
            ...
```

**Comment by ichorid** [comment]
Thanks, @WGH- ! It turned out we don't use `set_alert_notify` in Tribler. Also, I reproduced the bug with your code snippet, and the symptoms feel different from our problem described in #5319

**Comment by arvidn** [comment]
I've landed the `set_alert_fd()` patch in `RC_1_2`
